content/blog/custom-field-plugins.md

```diff
@@ -0,0 +1,144 @@
+
+There are **two steps to adding a field plugin to the CMS**. First, we'll define the field component object and register it with the CMS. Next, we'll use the field plugin in a form definition so we can edit content in the sidebar with our fancy custom field plugin.
+
+## 1. Add the Field Plugin to the CMS
+
+To register the custom field as a plugin with the CMS, we’ll need to head to the file where we can access the cms instance. In our Next.js demo, we’ll look at `_app.js`. This it the place where we override the default Next app to wrap `Tina` around each page.
+
+```js
+import React from 'react'
+import App from 'next/app'
+import { Tina, TinaCMS } from 'tinacms'
+import { GitClient } from '@tinacms/git-client'
+/*
+ ** 1. Import the custom field component
+ */
+import RangeInput from '../components/RangeInput'
+
+export default class Site extends App {
+  constructor() {
+    super()
+    this.cms = new TinaCMS({
+      apis: {
+        git: new GitClient('http://localhost:3000/___tina'),
+      },
+      sidebar: {
+        position: 'overlay',
+        hidden: process.env.NODE_ENV === 'production',
+      },
+    })
+
+    /*
+     ** 2. Define the field plugin
+     */
+    this.customFieldPlugin = {
+      name: 'range-input',
+      Component: RangeInput,
+    }
+    /*
+     ** 3. Register the plugin with the cms
+     */
+    this.cms.fields.add(this.customFieldPlugin)
+  }
+
+  render() {
+    const { Component, pageProps } = this.props
+    return (
+      <Tina cms={this.cms}>
+        <Component {...pageProps} />
+      </Tina>
+    )
+  }
+}
+```
+
+You’ll want to import the custom field component and then register the plugin with the CMS directly. Notice how we import the `RangeInput` component created in the [previous post](https://tinacms.org/blog/custom-field-components), this is the custom _field component_ that we're now attaching to a _field plugin_.
+
+> If you’re working with Gatsby, this [looks slightly different](https://tinacms.org/docs/gatsby/custom-fields/). _Hint_: you’ll head to the **gatsby-browser.js** file to access the CMS instance.
+
+### Field Plugin Definition
+
+A plugin is used to extend functionality in the CMS. A field plugin in particular allows us to create and register custom fields to the CMS.
+
+The field plugin definition **requires (at a minimum) a name and a component.** The name will be used to reference the custom field in form definitions. The component is what will actually be rendered in the sidebar.
+
+```ts
+interface FieldPlugin {
+  __type: 'field'
+  name: string
+  Component: React.FC<any>
+  type?: string
+  validate?(
+    value: any,
+    allValues: any,
+    meta: any,
+    field: Field
+  ): string | object | undefined
+  parse?: (value: any, name: string, field: Field) => any
+  format?: (value: any, name: string, field: Field) => any
+  defaultValue?: any
+}
+```
+
```

Comment:
It's not clear to me where to put that definition 😊

Reply:
I would switch from "definition" to "interface". What's described above is the interface for field plugins. You don't have to put it anywhere, it already exists in the tinacms codebase.